Fixing test step of Image Captioning example

badayvedat opened the pull request:
Test step of image captioning example is uses gradients of data and updates the weights accordingly. Model should not update its weights at test step. This causes model to fit on validation and test set.

fchollet commented:
@AakashKumarNain FYI

AakashKumarNain commented:
This looks like a blunder. I will take a look into it and fix it ASAP

fchollet commented:
@AakashKumarNain: could you please review the PR? You can always do a second pass on the example after the PR is merged.

AakashKumarNain commented:
@fchollet sure. There is only one thing that is required to change int the PR otherwise it is fine. Let me point out that in the code itself

AakashKumarNain commented on examples/vision/image_captioning.py:
```diff
@@ -409,38 +409,65 @@ def _compute_loss_and_acc(self, batch_data, training=True):
+                batch_seq_inp, encoder_out, training=False, mask=mask
+            )
+
+            # 5. Update loss and accuracy
```
Please change `#5 Update loss...` to `#6 Update loss ...`

AakashKumarNain commented on examples/vision/image_captioning.py:
```diff
@@ -409,38 +409,65 @@ def _compute_loss_and_acc(self, batch_data, training=True):
 
-            # 9. Get the gradients
+            # 8. Get the gradients
             grads = tape.gradient(loss, train_vars)
```
The update here is for each caption. Either keep the `batch_loss` where you accumulate the loss for each caption or just use another name for per caption loss

badayvedat replied:
I think using different variable names would be more appropriate. I've changed the variable names to clearly distinguish between the losses.

badayvedat commented on examples/vision/image_captioning.py:
```diff
@@ -461,12 +465,16 @@ def test_step(self, batch_data):
-        loss = loss / float(self.num_captions_per_image)
-        acc = acc / float(self.num_captions_per_image)
+        loss = batch_loss
+        acc = batch_acc / float(self.num_captions_per_image)
```
In earlier commit, the loss was divided by the number of captions. Since we're interested in accumulative loss, I removed the dividing part. But I'm not so sure which one would be more convenient. Let me know what you think about it.

fchollet reviewed (commented):
Thank you! Please add the generated files. (Command `python autogen.py add_example vision/image_captioning`)

badayvedat commented:
@fchollet When I call the autogen.py I get the following error:
`keras-io/scripts/tutobooks.py", line 176, in py_to_nb
    raise ValueError(
ValueError: Found 328 lines of code, but expected fewer than 300
`
MAX_LOC value is defined as 300 in keras-io/scripts/tutobooks.py, line 72.
Is there a workaround for it without changing the MAX_LOC value?  I already checked the code and I was not able to decrease loc.

fchollet commented:
> Is there a workaround for it without changing the MAX_LOC value? I already checked the code and I was not able to decrease loc.

You can change the MAX_LOC value in the file directly. Just don't push that change.

badayvedat commented:
I've added the generated files. Let me know if there's something else I need to do.

In addition to that, since I fixed the test step, accuracy decreased drastically. I did a little hyperparameter tuning(changing learning rate) but accuracy is still low. Would be good if someone does a hyper parameter search in the future.

badayvedat commented:
@fchollet any updates on this?

fchollet reviewed (approved):
Thanks for the update, LGTM
